### ISSUE TITLE
Deno.readDir won't return all files

### ISSUE THREAD

**Issue by Sembiance** [denoland/deno#12998]
If a filename has funny characters `Deno.readDir` will silently ignore those files.

First unzip `bug.tar.zip` which produces `bug.tar` file
Now untar `bug.tar` which produces `bug` directory
Now run `deno run --allow-read readDirBug.js`:

```js
for await(const entry of Deno.readDir("bug")) {
	console.log(entry);
}
```

It only shows 2 entries, should be 4. 

NodeJS finds these files just fine:
```js
import { readdir } from 'fs/promises';

const entries = await readdir("bug");
for (const entry of entries) {
	console.log(entry);
}
```

**Comment by lucacasonato** [comment]
Cause: https://github.com/denoland/deno/blob/d31378726e78490e88b8a9ec3001b86ea009d978/runtime/ops/fs.rs

**Comment by andreubotella** [comment]
> Cause: https://github.com/denoland/deno/blob/d31378726e78490e88b8a9ec3001b86ea009d978/runtime/ops/fs.rs

You didn't link to a specific line – I'm assuming you mean

https://github.com/denoland/deno/blob/d31378726e78490e88b8a9ec3001b86ea009d978/runtime/ops/fs.rs#L1176

----------------

I've done some work on puzzling out the web-exposed filenames that you get from file controls, at w3c/FileAPI#161. Of course, this gets a lot more complicated with file system access APIs. Is this something we want?

**Comment by lucacasonato** [comment]
> You didn't link to a specific line – I'm assuming you mean

Yeah. Github's "Copy permalink" doesn't copy line numbers it seems 😔

I think the behavior we want is whatever Node does here. So either replace the invalid chars with replacement chars, or use WTF-16 encoding.

**Comment by Sembiance** [comment]
Forgot to attach the bug.tar.zip [bug.tar.zip](https://github.com/denoland/deno/files/7660779/bug.tar.zip)

**Comment by Sembiance** [comment]
Looks like this is a long standing issue: https://github.com/denoland/deno/issues/627

Sounds like one possibly solution would be a Deno equilivant `OsString` that could be used Deno wide with all file operations including readDir. However based on the discussion in that issue, that would be a huge undertaking.

**Comment by shanmukhateja** [comment]
Hi @Sembiance @lucacasonato I would like to take a shot at this.

Do we have a public list of replacement chars to use? I'd like to try the "replace invalid chars" approach.

**Comment by andreubotella** [comment]
Hi @shanmukhateja, and thank you for your interest, but unfortunately this isn't a simple issue of replacing characters from the JS side. The code that deals with interacting with the filesystem is Rust code, including the code that skips files with invalid filenames here. And changing it to not skip them would already be enough to replace those invalid bytes or code units with the replacement character (U+FFFD indeed).

But making this change would have further effects on the Deno ecosystem, because if only `Deno.readDir()` is changed and not the rest of the filesystem APIs, it would mean that some of the files returned by `Deno.readDir()` would not be accessible with `Deno.readTextFile()` or other APIs, because you could have two files with different filenames that map to the same Deno-exposed filename.

**Comment by bartlomieju** [comment]
Fixed in #34623 ("fix(ext/fs): surface non-UTF-8 file names from read_dir").

`Deno.readDir` no longer silently drops entries with non-UTF-8 names. They're now decoded lossily (invalid bytes become U+FFFD) so they appear in directory listings, matching Node's default utf8 `readdir` behavior — which is what this thread converged on.

Note that the deeper concern raised in the discussion — that a lossily-decoded name may not round-trip back through `Deno.readTextFile()` and friends, or that two distinct files could collide on the same U+FFFD-containing name — is not addressed by this change. That would require an `OsString`-style filesystem API and is tracked in #627.

Closing this one since the reported bug (readDir not returning all files) is resolved.

**Comment by Sembiance** [comment]
I ended up long ago running convmv recursively to ensure names are UTF8 before attempting to list/access the files with deno:
`convmv -r --qfrom --qto --notest -f windows-1252 -t UTF-8 /path/to/parent/dir`

By the way, great to see all this movement/activity on longstanding issues! :)
